### ISSUE TITLE
v2.1.0 Testers Needed - MQTT support for Motion Detect!

### ISSUE THREAD

**Issue by donavanbecker** [homebridge-plugins/homebridge-camera-ffmpeg#575]
### Changes

* Add MQTT support for Motion Detect (#572), thanks to [fennec622](https://github.com/fennec622).
* Just publish to topic, no message, `homebridge/motion/Camera_name`, You must add `_`  to your camera if you have space in the name.

```
{
  "platform": "Camera-ffmpeg",
  "mqtt": "127.0.0.1",
  "topics": "homebridge/motion/#",
  "cameras": [
    {
      "name": "Camera name",
      "motion": true,
      "videoConfig": {
        "source": "-re -i http://10.0.0.1",
        "stillImageSource": "-i http://10.0.0.1",
        "maxStreams": 5,
        "maxWidth": 1280,
        "maxHeight": 720,
        "maxFPS": 15,
        "maxBitrate": 1000
      }
    }
  ]
}
```

**Comment by donavanbecker** [comment]
You can test this by installing latest beta
* `sudo npm install -g --unsafe-perm homebridge-camera-ffmpeg@beta`
* To Revert `sudo npm install -g --unsafe-perm homebridge-camera-ffmpeg@latest`

**Comment by isramos** [comment]
I've been testing over the last 24 hours, and seem to be working fine.

Some notes:
- the installation method above did not work for me. Instead, I went to HOOBS terminal and `cd /home/hoobs/.hoobs`, where package.json is located. Then ran `sudo npm install -g --unsafe-perm homebridge-camera-ffmpeg@beta` 
- mqtt topic: I wish there was an option to define the camera topic in `cameras[].topic` as an alternative to `cameras[].name`.
 - I wonder why do we still need the dummy toggle switch.
-  Config enter in the HOOBS pretty UI is not being picked up. I wonder if `config.schema.json` line 24, should say `topics` instead of `topic`, so the UI parameter matches the actual property being picked up by code. I'm just guessing here.

Thanks for the great work!

**Comment by donavanbecker** [comment]
> Thanks for the great work!

Thanks for the feedback!

**Comment by donavanbecker** [comment]
@isramos we have added an option to remove the dummy toggle switches and also added stateless doorbell switch for automations. If you could try that out?

**Comment by donavanbecker** [comment]
please test out latest beta. homebridge-camera-ffmpeg@2.0.2-beta.37`

**Comment by tom301080** [comment]
Still having the issue with ffmpeg exiting even with beta39

<img width="1511" alt="Bildschirmfoto 2020-07-03 um 18 02 59" src="https://user-images.githubusercontent.com/48982934/86484911-99e5cc80-bd57-11ea-9dc8-f0a9db7d711e.png">

**Comment by california444** [comment]
Hi,

some users mentioned in  #554 that they can not restart homebridge due to an issue with multiple services with same UUID, e.g.[link](https://github.com/homebridge-plugins/homebridge-camera-ffmpeg/issues/554#issuecomment-653178948)

The stacks trace in the referenced link shows that the DoorbellSwitch service in line 123 is the issue.
This service is not removed before like all other accessory services and thus could exist multiple times.
https://github.com/homebridge-plugins/homebridge-camera-ffmpeg/blob/94cfc89187e1ff2c7150e056cc78474ce1fedf8e/src/index.ts#L121-L128


Could it be required to remove this service as well as all the other doorbell, motion and switch services? 
https://github.com/homebridge-plugins/homebridge-camera-ffmpeg/blob/94cfc89187e1ff2c7150e056cc78474ce1fedf8e/src/index.ts#L77-L93

**Comment by donavanbecker** [comment]
@california444 a pull request for this fix would be welcomed.

**Comment by isramos** [comment]
I just tested the beta tag. Here are my notes:

Whenever motion is detected I send MQTT message:
- topic: `/homebridge/motion/Camera One`
- payload: `{"MotionDetected":true}`

I see in the logs `MQTT state message received: {"MotionDetected":true}`. I conclude that we're now looking for the camera name in the payload. Maybe there should be a message in the logs saying that the camera called `{"MotionDetected":true}` is not found the list of configured cameras: [ 'Camera One', 'Camera Two']. That would make my mistake more obvious. 

Once I transmit payload: `Camera One` I see the motion detection triggering: `Switch Motion Detect On : Camera One`. 

**IMPORTANT:**  However, I don't see it self clearing. It stays triggered forever. Please test it **IMPORTANT:**

One thing to consider: I suggest decoupling the camera name and use a different property to matching. Let's say I have a portable camera that each day I move to a different room, and each time I give it a different name. It would be a hassle to keep things in sync. On the other hand, if we use, say the serialNumber of the camera (or other prop) that would make it a more robust approach.
``
`
                    "model": "TBD",
                    "serialNumber": "0001",
                    "manufacturer": "LUX",
                    "name": "Cam Driveway",
   
`` `

In other words: topic would be: `/homebridge/motion/{ serialNumber || name}`

**Comment by DJBenson** [comment]
> Still having the issue with ffmpeg exiting even with beta39
> 
> <img alt="Bildschirmfoto 2020-07-03 um 18 02 59" width="1511" src="https://user-images.githubusercontent.com/48982934/86484911-99e5cc80-bd57-11ea-9dc8-f0a9db7d711e.png">

Same here :(

Totally "clean" Homebridge install, latest beta of the plugin, added a camera with a completely different name, no joy.

**Comment by ex725** [comment]
what settings do I need to make in these files to test
![image](https://user-images.githubusercontent.com/56446241/86498280-daf8d380-bd8d-11ea-9e82-7d81be874b9b.png)
![image](https://user-images.githubusercontent.com/56446241/86498316-02e83700-bd8e-11ea-83f4-4b93e2597fbb.png)

**Comment by burnbrigther** [comment]
Also still getting "[FFMPEG] ffmpeg exited with code: 1 and signal: null (error)".  This is my last camera in my configuration and the only one having this problem.

`[7/3/2020, 3:27:56 PM] [Camera-ffmpeg] Starting FRONT DOOR (Axis) video stream (1280x720, 10 fps, 299 kbps, 1316 mtu)...`
`[7/3/2020, 3:27:56 PM] [Camera-ffmpeg] [FFMPEG] ffmpeg exited with code: 1 and signal: null (error)`
`[7/3/2020, 3:27:56 PM] [Camera-ffmpeg] Stopped FRONT DOOR (Axis) video stream!`

**Comment by steverae68** [comment]
OK so call me dumb, but can someone point me in the direction or indeed, explain what the point of this MQTT addition is to this plug.   At present i have several camera's configured, these are all Hik-Vision POE cams linked to a central NVR.

These cams all have built in motion detection and report to a App with video clips when motion is detected as you would expect.

Now using the v2.0.1 edition of this plugin, i have a Philips motion detector device which detects motion and switches on the camera dummy switch (via a homekit automation) which in turn triggers the motion sensor part of this plugin and generates the notification to Homekit itself.

If the idea that this MQTT will replace the need for the dummy switch and be able to trigger from the camera's own motion sensor (not sure if this is possible).

I'd just like a quick brief explanation please......

**Comment by Sunoo** [comment]
Okay, I can't for the life of me figure out how to get this to work.

My camera (a WyzeCam2 running Dafang Hacks) publishes to the topic configured followed by '/motion' with the word 'ON' or 'OFF' depending on whether motion is detected.

The name of my camera in this plugin is "Cat Food Camera" so I've tried configuring the camera to publish to "homebridge/motion/Cat Food Camera" and "homebridge/motion/Cat_Food_Camera" and in both cases I'm able to see messages coming (as described above) through when I subscribe to the topics myself. However, nothing I've tried has been able to get this plugin to show that it's seeing anything, though it does log "[Camera-ffmpeg] MQTT CONNECTED!" when restarting Homebridge, I never see anything else MQTT-related, and the motion sensor never changes states.

**Comment by Sunoo** [comment]
Just a thought, but as someone else mentioned, perhaps defining the MQTT topic for each camera would be a better approach, and it’d certainly be more clear when configuring.

Some way to specify the messages for motion detected or not (perhaps regex?) might be a useful addition as well, since that likely varies from device to device.

**Comment by donavanbecker** [comment]
@fennec622

**Comment by Sunoo** [comment]
So I grabbed the code and started poking around, and it works differently than the documentation at the head of the thread claims.

> Just publish to topic, no message, homebridge/motion/Camera_name, You must add _ to your camera if you have space in the name.

The way it _actually_ seems to work, is that you need to publish `Camera name` (without the spaces replaced with underscores) as a message to homebridge/motion.

The code as currently written also only swaps the first underscore with a space. So in my case with the camera named 'Cat Food Camera', if I follow the directions and use 'Cat_Food_Camera', the plugin converts it to 'Cat Food_Camera', which means the string comparison never works. However, ignoring that instruction and publishing the message of 'Cat Food Camera' causes the motion sensor to trip.

**Comment by fennec622** [comment]
Just publish message name of camera 

To topic You choose 

More simply to specify one topic for all cameras 

And adjust message for specify camera

I see for replace all Space not only the  first

**Comment by Sunoo** [comment]
@fennec622 In Javascript string.replace only works on the first instance of the pattern unless regular expressions are used.

I just created pull request #604 to correct this issue. Though, I don't see the need to do this replacement at all, since spaces are a valid character in MQTT messages?

I'm also still not convinced that relying on a single topic to handle everything is the best way forward, as that basically mandates some kind of shim program in most cases to subscribe to the MQTT topic that the camera actually publishes to, then republish it in the format this plugin now expects.

**Comment by Sunoo** [comment]
Additionally, I recommend updating the first post as such, to reflect the way this plugin is currently actually working:

> ## v2.X.X (2020-XX-XX)
> ### Changes
> * Add MQTT support for Motion Detect (#572), thanks to [fennec622](https://github.com/fennec622).
> * Just publish to topic with the camera name as the message, `homebridge/motion`. You may use `_`  in the camera name instead of a space if you choose.
> 
> ```
> {
>   "platform": "Camera-ffmpeg",
>   "mqtt": "127.0.0.1",
>   "topics": "homebridge/motion",
>   "cameras": [
>     {
>       "name": "Camera name",
>       "motion": true,
>       "videoConfig": {
>         "source": "-re -i http://10.0.0.1",
>         "stillImageSource": "-i http://10.0.0.1",
>         "maxStreams": 5,
>         "maxWidth": 1280,
>         "maxHeight": 720,
>         "maxFPS": 15,
>         "maxBitrate": 1000
>       }
>     }
>   ]
> }
> ```

**Comment by fennec622** [comment]
more simply i make change
just post message with same name of camera
post message

Cat Food Camera

on topic homebridge/motion

**Comment by Sunoo** [comment]
@fennec622 I agree that this is the less confusing approach. I've updated my pull request to strip out that replacement instead. That would mean the first post should be changed to:


> ## v2.X.X (2020-XX-XX)
> ### Changes
> * Add MQTT support for Motion Detect (#572), thanks to [fennec622](https://github.com/fennec622).
> * Just publish to topic `homebridge/motion` with the camera name as the message.
> 
> ```
> {
>   "platform": "Camera-ffmpeg",
>   "mqtt": "127.0.0.1",
>   "topics": "homebridge/motion",
>   "cameras": [
>     {
>       "name": "Camera name",
>       "motion": true,
>       "videoConfig": {
>         "source": "-re -i http://10.0.0.1",
>         "stillImageSource": "-i http://10.0.0.1",
>         "maxStreams": 5,
>         "maxWidth": 1280,
>         "maxHeight": 720,
>         "maxFPS": 15,
>         "maxBitrate": 1000
>       }
>     }
>   ]
> }
> ```

Also, I still think it would be nice if, at minimum, there was a way to have two different messages to trigger or clear the motion sensor. It could be set up so that if there is no 'clear' message defined (maybe if motion sensor timeout value is set to '0' in the config or similar?) then the timeout fires and clears the the sensor automatically.

**Comment by fennec622** [comment]
For the moment reset motion is timeout to 1s

But it's possible to add config timer to reset motion for all cameras 

"motionreset": 1;

**Comment by Sunoo** [comment]
@fennec622 No reason to increase time, but I thought that making it configurable gives an easy spot to disable the reset.

The reason for that is that my camera sends messages for both when it sees motion and when it stops seeing motion. It would be nice to have a way to support both of those instead of blinding resetting things, so that the motion sensor in HomeKit matches the motion detected state of the camera.

This would make the motion sensor useful beyond just having tripping send a push, I could look at the sensor in Home and see whether motion is currently being detected or not.

**Comment by fennec622** [comment]
Ok
I see that for add reset motion by mqtt

**Comment by Sunoo** [comment]
Yes, that's exactly what I was getting at.

**Comment by Sunoo** [comment]
Okay, so I figured out a fairly clean way of supporting the way my camera publishes to MQTT. I made another Homebridge plugin (https://github.com/Sunoo/homebridge-dafang-mqtt-republish) that just subscribes to what my camera sends and republishes it in the way this plugin expects. And it works quite well.

It would still be very nice to be able to have an MQTT message that resets the motion sensor instead of using the timeout though. I could very easily update my plugin to support that, if some method to do so is added to this plugin.

Thanks.

**Comment by donavanbecker** [comment]
closing due to v2.1.0 being release.

**Comment by california444** [comment]
@donavanbecker v2.1.0 works like a charm here, thanks a lot!!

**Comment by donavanbecker** [comment]
@california444 thanks for your input!!!

**Comment by steverae68** [comment]
> OK so call me dumb, but can someone point me in the direction or indeed, explain what the point of this MQTT addition is to this plugin.
> At present i have several camera's configured, these are all Hik-Vision POE cams linked to a central NVR.
> 
> These cams all have built in motion detection and report to a App with video clips when motion is detected as you would expect.
> 
> Now using the v2.0.1 edition of this plugin, i have a Philips motion detector device which detects motion and switches on the camera dummy switch (via a homekit automation) which in turn triggers the motion sensor part of this plugin and generates the notification to Homekit itself.
> 
> If the idea that this MQTT will replace the need for the dummy switch and be able to trigger from the camera's own motion sensor (not sure if this is possible).
> 
> I'd just like a quick brief explanation please...... what would be good would be a way to capture the notification sent from the the Hik-Connect app (or which ever app is generating the notification) and use that as a trigger

Can anyone provide a step by step guide to set this up?  ive installed the plugin homebridge-mqttthing but im now at a loss

If it is not too much to ask - my email is steve.rae@sky.com

thanks
